staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go

```diff
@@ -0,0 +1,148 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Watcher is an example of watching on resource create/update/delete events
+type Watcher struct {
+	clientset     *kubernetes.Clientset
+	exampleClient *rest.RESTClient
+	exampleScheme *runtime.Scheme
+}
+
+// Run starts an Example resource watcher
+func (w *Watcher) Run(ctx context.Context) error {
+	fmt.Printf("Watch Example objects\n")
+
+	// Watch Example objects
+	handler := ExampleEventHandler{}
+	exampleInformer, err := watchExamples(ctx, w.exampleClient, w.exampleScheme, &handler)
+	if err != nil {
+		fmt.Printf("Failed to register watch for Example resource: %v\n", err)
+		return err
+	}
+
+	fmt.Printf("WaitForCacheSync\n")
+	// We must wait for exampleInformer to populate its cache to avoid reading from an empty cache
+	// in case of resource-generated evxents.
+	if !cache.WaitForCacheSync(ctx.Done(), exampleInformer.HasSynced) {
+		return errors.New("Wait for Example Informer was cancelled")
+	}
+
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func watchExamples(ctx context.Context, exampleClient cache.Getter, exampleScheme *runtime.Scheme, handler cache.ResourceEventHandler) (cache.Controller, error) {
+	parameterCodec := runtime.NewParameterCodec(exampleScheme)
+
+	source := newListWatchFromClient(
+		exampleClient,
+		ExampleResourcePath,
+		api.NamespaceAll,
+		fields.Everything(),
+		parameterCodec)
+
+	store, controller := cache.NewInformer(
+		source,
+
+		// The object type.
+		&Example{},
+
+		// resyncPeriod
+		// Every resyncPeriod, all resources in the cache will retrigger events.
+		// Set to 0 to disable the resync.
+		0,
+
+		// Your custom resource event handlers.
+		handler)
+
+	// store can be used to List and Get
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	for _, obj := range store.List() {
+		example := obj.(*Example)
+
+		// This will likely be empty the first run, but may not
+		fmt.Printf("Existing example: %#v\n", example)
+	}
+
+	go controller.Run(ctx.Done())
+
+	return controller, nil
+}
+
+// See the issue comment: https://github.com/kubernetes/kubernetes/issues/16376#issuecomment-272167794
+// newListWatchFromClient is a copy of cache.NewListWatchFromClient() method with custom codec
+// Cannot use cache.NewListWatchFromClient() because it uses global api.ParameterCodec which uses global
+// api.Scheme which does not know about custom types (Example in our case) group/version.
```

Comment:
As far as I see, the issue is not that the custom types `Example` etc. are not registered, but that `ListOptions` and friends must be registered under the `config.GroupVersion` of the client. Compare the registration in https://github.com/kubernetes/kubernetes/blob/db0b0bd0f5f934ef83cce0bd35212aba5bb31e18/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/register.go#L46 and the conversion at https://github.com/kubernetes/kubernetes/blob/fb78390f839cd1241ef8cf8e4b257447001ba962/staging/src/k8s.io/apimachinery/pkg/runtime/codec.go#L188.

In fact, I think we should do a proper comparison with all gvks at https://github.com/kubernetes/kubernetes/blob/fb78390f839cd1241ef8cf8e4b257447001ba962/staging/src/k8s.io/apimachinery/pkg/runtime/codec.go#L187. After all the order of registration returned by `ObjectKinds` are arbitrary and uncontroller. @smarterclayton wdyt? Obviously, this is nothing for this PR.